### PR TITLE
Support  sku related node event

### PR DESCRIPTION
**Pull request by anhou**
require https://github.com/RackHD/on-core/pull/178 asynchronous publish event when generating sku.  docs: https://github.com/RackHD/docs/pull/296

@RackHD/corecommitters @zyoung51  @WangWinson

### CODE REVIEW

**Comment by WangWinson** [comment]
:+1:

**Comment by anhou** [comment]
NOTE: dependency exists, need to merge https://github.com/RackHD/on-core/pull/178 firstly

File: lib/jobs/generate-sku.js [hunk 1]
```diff
@@ -118,13 +118,19 @@ function generateSkuJobFactory(
...
-                    if (match && match.sku) {
-                        return eventsProtocol.publishSkuAssigned(self.nodeId, match.sku.id);
-                    }
+                return waterline.nodes.needByIdentifier(self.nodeId)
+                .then(function(oldNode) {
+                    return waterline.nodes.updateByIdentifier(self.nodeId, {
+                        sku: match && match.sku ? match.sku.id : null
+                    })
+                    .then(function(newNode) {
+                        /* asynchronous */
```
Comment (yyscamper) [inline lib/jobs/generate-sku.js]: Could you please explain more in the code you intentionally want the promise not be chained but run in background.

File: lib/jobs/generate-sku.js [hunk 2]
```diff
@@ -118,13 +118,19 @@ function generateSkuJobFactory(
...
+                        eventsProtocol.publishNodeAttrEvent(oldNode, newNode, 'sku');
+
+                        if (match && match.sku) {
+                            return eventsProtocol.publishSkuAssigned(self.nodeId, match.sku.id);
```
Comment (yyscamper) [inline lib/jobs/generate-sku.js]: You intentionally not return previous promise, I think it's better not to return this as well

File: lib/jobs/generate-sku.js [hunk 3]
```diff
@@ -118,13 +118,19 @@ function generateSkuJobFactory(
...
-                        return eventsProtocol.publishSkuAssigned(self.nodeId, match.sku.id);
-                    }
+                return waterline.nodes.needByIdentifier(self.nodeId)
+                .then(function(oldNode) {
+                    return waterline.nodes.updateByIdentifier(self.nodeId, {
+                        sku: match && match.sku ? match.sku.id : null
+                    })
+                    .then(function(newNode) {
+                        /* asynchronous */
+                        eventsProtocol.publishNodeAttrEvent(oldNode, newNode, 'sku');
```
Comment (yyscamper) [inline lib/jobs/generate-sku.js]: Add `catch` to the promise

File: lib/jobs/generate-sku.js (hunk 2 above)
Comment (anhou) [inline lib/jobs/generate-sku.js]: it's previous code,  not intend to change it, and maybe refined to remove this function in near futre

**Comment by yyscamper** [comment]
:+1: after my comment is addressed

**Comment by jlongever** [comment]
test this please

**Comment by anhou** [comment]
test this please

**Pull request merged**